metrics-core/src/lib.rs

```diff
@@ -59,3 +61,28 @@ pub trait MetricsRecorder {
     /// of the underlying observed values, and callers will need to process them accordingly.
     fn record_histogram<K: AsRef<str>>(&mut self, key: K, values: &[u64]);
 }
+
+/// A value that holds a point-in-time view of collected metrics.
+pub trait Snapshot {
+    /// Records the snapshot to the given recorder.
+    fn record<R: Recorder>(&self, recorder: &mut R);
+}
+
+/// A value that can provide on-demand snapshots.
+pub trait SnapshotProvider {
+    type Snapshot: Snapshot;
+    type SnapshotError: Display;
```

Comment:
`: Error` bound instead? or do you even need to bound this?

Reply:
So, for the non-async stuff, it has to have a bound so we can display it.  `Error` could probably work fine here, since it requires `Display` and `Debug` to begin with.

But yeah, if we don't bound it, then we can't log the error.

Reply:
Sorry, that's to say.... we don't need to bound it necessarily..  I suppose we could defer the bound to the users -- i.e. the exporters -- who make the decision on whether or not they're going to log/emit/etc any errors they get.

I'll move the bound there, and change `SnapshotError` (the actual struct type, not the associated type in the trait) to impl `Error` and then bound on `Error` for the log exporter, just to show the proper way to do it for future exporter impls.

Reply:
Yeah, this bound should probably be moved to who ever takes this impl.

Reply:
Yep, makes sense.  Will do.